src/Grpc.AspNetCore.Web/Internal/Base64PipeReader.cs

```diff
@@ -0,0 +1,170 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Buffers;
+using System.Buffers.Text;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Grpc.AspNetCore.Web.Internal
+{
+    /// <summary>
+    /// Reads and decodes base64 encoded bytes from the inner reader.
+    /// </summary>
+    internal class Base64PipeReader : PipeReader
+    {
+        private readonly PipeReader _inner;
+        private ReadOnlySequence<byte> _currentInnerBuffer;
+        private ReadOnlySequence<byte> _currentDecodedBuffer;
+
+        public Base64PipeReader(PipeReader inner)
+        {
+            _inner = inner;
+        }
+
+        public override void AdvanceTo(SequencePosition consumed)
+        {
+            var consumedPosition = ResolvePosition(consumed);
+
+            _inner.AdvanceTo(consumedPosition);
+        }
+
+        public override void AdvanceTo(SequencePosition consumed, SequencePosition examined)
+        {
+            var consumedPosition = ResolvePosition(consumed);
+            var examinedPosition = ResolvePosition(examined);
+
+            _inner.AdvanceTo(consumedPosition, examinedPosition);
+        }
+
+        private SequencePosition ResolvePosition(SequencePosition base64Position)
+        {
+            // TODO - is there a better way to get the index of a position in a sequence?
+            var length = _currentDecodedBuffer.Slice(0, base64Position).Length;
+
+            var resolvedPosition = _currentInnerBuffer.GetPosition(Base64.GetMaxEncodedToUtf8Length((int)length));
```

Comment:
Could GetMaxEncodedToUtf8Length ever overestimate the number of bytes that were consumed in the _currentInnerBuffer because the the base 64 encoding happened to be more efficient than it could have been given a different payload of the same length? Or is Base64.GetMaxEncodedToUtf8Length guaranteed to be exact? That latter seems unlikely to me given the name.

Reply:
It will divide by 3 (with roundup for base64 padding) then multiply by 4.

You're right that this could potentially resolve to a higher number because of the roundup, but in practice it won't happen because of how gRPC uses the PipeReader. ReadAsync will always return bytes in multiples of 3 (until the very end of the message), and PipeReader.Advance will always be called with ReadResult.Buffer.End. The only time roundup will happen is at the end of the message which is where we want the roundup.

I'll add a note that this is an assumption/limitation of the reader.